### PR TITLE
[FLINK-11974][runtime] Introduce StreamOperatorFactory

### PR DESCRIPTION
**Pull request by JingsongLi**
## What is the purpose of the change

Introduce StreamOperatorFactory to help table perform the whole Operator CodeGeneration

## Brief change log

Introduce StreamOperatorFactory
Introduce SimpleOperatorFactory
Introduce SetupableStreamOperator to remove setup in StreamOperator

## Verifying this change

ut

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (no)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (yes)
  - The serializers: (no)
  - The runtime per-record code paths (performance sensitive): (no)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Yarn/Mesos, ZooKeeper: (no)
  - The S3 file system connector: (no)

## Documentation

  - Does this pull request introduce a new feature? (yes)
  - If yes, how is the feature documented? (JavaDocs)

### CODE REVIEW

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/api/graph/StreamGraphGenerator.java [hunk 1]
```diff
@@ -480,13 +483,17 @@ private StreamGraph generateInternal(List<StreamTransformation<?>> transformatio
...
 				"Source: " + source.getName());
-		if (source.getOperator().getUserFunction() instanceof InputFormatSourceFunction) {
-			InputFormatSourceFunction<T> fs = (InputFormatSourceFunction<T>) source.getOperator().getUserFunction();
-			streamGraph.setInputFormat(source.getId(), fs.getFormat());
+		if (source.getOperatorFactory() instanceof SimpleOperatorFactory) {
+			SimpleOperatorFactory<T> factory = (SimpleOperatorFactory<T>) source.getOperatorFactory();
+			Function userFunction = ((StreamSource) factory.getOperator()).getUserFunction();
+			if (userFunction instanceof InputFormatSourceFunction) {
+				InputFormatSourceFunction<T> fs = (InputFormatSourceFunction<T>) userFunction;
+				streamGraph.setInputFormat(source.getId(), fs.getFormat());
```
Comment (JingsongLi) [inline flink-streaming-java/src/main/java/org/apache/flink/streaming/api/graph/StreamGraphGenerator.java]: Here need get input format, not exposing this capability to StreamOperatorFactory

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/api/graph/StreamingJobGraphGenerator.java [hunk 2]
```diff
@@ -648,12 +649,14 @@ private void configureCheckpointing() {
...
-				Function f = ((AbstractUdfStreamOperator<?, ?>) op).getUserFunction();
-
-				if (f instanceof WithMasterCheckpointHook) {
-					hooks.add(new FunctionMasterCheckpointHookFactory((WithMasterCheckpointHook<?>) f));
+			if (node.getOperatorFactory() instanceof SimpleOperatorFactory) {
+				SimpleOperatorFactory factory = (SimpleOperatorFactory) node.getOperatorFactory();
+				if (factory.getOperator() instanceof AbstractUdfStreamOperator) {
+					Function f = ((AbstractUdfStreamOperator<?, ?>) factory.getOperator()).getUserFunction();
+
+					if (f instanceof WithMasterCheckpointHook) {
```
Comment (JingsongLi) [inline flink-streaming-java/src/main/java/org/apache/flink/streaming/api/graph/StreamingJobGraphGenerator.java]: not exposing this capability to StreamOperatorFactory too.

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/api/transformations/OneInputTransformation.java [hunk 3]
```diff
@@ -82,11 +94,16 @@ public OneInputTransformation(
 		return input.getOutputType();
 	}
 
+	@VisibleForTesting
+	public OneInputStreamOperator<IN, OUT> getOperator() {
```
Comment (JingsongLi) [inline flink-streaming-java/src/main/java/org/apache/flink/streaming/api/transformations/OneInputTransformation.java]: I'm not sure whether to keep it or to modify all the tests. Same to others.

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/api/graph/JSONGenerator.java [hunk 4]
```diff
@@ -187,8 +186,6 @@ private void decorateNode(Integer vertexID, ObjectNode node) {
 			node.put(PACT, "Operator");
 		}
 
-		StreamOperator<?> operator = streamGraph.getStreamNode(vertexID).getOperator();
```
Comment (pnowojski) [inline flink-streaming-java/src/main/java/org/apache/flink/streaming/api/graph/JSONGenerator.java]: Is this change relevant to the rest of the PR? If not, could you extract changes in this file to a separate `[hotfix]` commit?

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/api/graph/StreamGraphGenerator.java [hunk 5]
```diff
@@ -480,13 +483,17 @@ private StreamGraph generateInternal(List<StreamTransformation<?>> transformatio
...
 				source.getCoLocationGroupKey(),
-				source.getOperator(),
+				source.getOperatorFactory(),
 				null,
 				source.getOutputType(),
 				"Source: " + source.getName());
-		if (source.getOperator().getUserFunction() instanceof InputFormatSourceFunction) {
-			InputFormatSourceFunction<T> fs = (InputFormatSourceFunction<T>) source.getOperator().getUserFunction();
-			streamGraph.setInputFormat(source.getId(), fs.getFormat());
+		if (source.getOperatorFactory() instanceof SimpleOperatorFactory) {
```
Comment (pnowojski) [inline flink-streaming-java/src/main/java/org/apache/flink/streaming/api/graph/StreamGraphGenerator.java]: Possible issue: this code doesn't support setting input formats for non `SimpleOperatorFactories`. By implementing this check this way, we are supporting all of the present cases, but it makes it kind of strange, that for the new way we don't support it - we do not have a migration path to get rid of `StreamOperators` from the `StreamTransformation`.

Could this if check be reworked to something like:
```
if (source.getOperatorFactory() insnaceof InputFormatSourceOperatorFactory) {
  streamGraph.setInputFormat(id, ((InputFormatSourceOperator) source.getOperatorFactory()).getFormat());
}
```
?
Combination of `SimpleOperatorFactory` and `InputFormatSourceOperatorFactory` could implement `getFormat() { return this.getOperator().getUserFunction().getFormat()`.

while future non `SimpleOperatorFactory` could be supported as well.

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/api/graph/StreamGraphHasherV2.java [hunk 6]
```diff
@@ -252,9 +253,12 @@ private boolean generateNodeHash(
 
 		if (LOG.isDebugEnabled()) {
 			String udfClassName = "";
-			if (node.getOperator() instanceof AbstractUdfStreamOperator) {
-				udfClassName = ((AbstractUdfStreamOperator<?, ?>) node.getOperator())
-						.getUserFunction().getClass().getName();
+			if (node.getOperatorFactory() instanceof SimpleOperatorFactory) {
```
Comment (pnowojski) [inline flink-streaming-java/src/main/java/org/apache/flink/streaming/api/graph/StreamGraphHasherV2.java]: ditto about factories. `if (node.getOperatorFactory() instanceof UdfStreamOperatorFactory)`?

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/api/graph/StreamingJobGraphGenerator.java [hunk 7]
```diff
@@ -648,12 +649,14 @@ private void configureCheckpointing() {
 		final ArrayList<MasterTriggerRestoreHook.Factory> hooks = new ArrayList<>();
 
 		for (StreamNode node : streamGraph.getStreamNodes()) {
-			StreamOperator<?> op = node.getOperator();
-			if (op instanceof AbstractUdfStreamOperator) {
-				Function f = ((AbstractUdfStreamOperator<?, ?>) op).getUserFunction();
-
-				if (f instanceof WithMasterCheckpointHook) {
-					hooks.add(new FunctionMasterCheckpointHookFactory((WithMasterCheckpointHook<?>) f));
+			if (node.getOperatorFactory() instanceof SimpleOperatorFactory) {
```
Comment (pnowojski) [inline flink-streaming-java/src/main/java/org/apache/flink/streaming/api/graph/StreamingJobGraphGenerator.java]: ditto `if (node.getOperatorFactory() instanceof UdfStreamOperatorFactory)`?

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/api/operators/StreamOperatorFactory.java [hunk 8]
```diff
@@ -0,0 +1,87 @@
...
+import org.apache.flink.streaming.runtime.streamrecord.StreamRecord;
+import org.apache.flink.streaming.runtime.tasks.StreamTask;
+
+import java.io.Serializable;
+
+/**
+ * A factory to create {@link StreamOperator}.
+ *
+ * @param <OUT> The output type of the operator
+ */
```
Comment (pnowojski) [inline flink-streaming-java/src/main/java/org/apache/flink/streaming/api/operators/StreamOperatorFactory.java]: `@PublicEvolving`? `@Internal`? (question to someone from API team)

File: flink-table/flink-table-runtime-blink/src/main/java/org/apache/flink/table/runtime/OneInputOperatorWrapper.java [hunk 9]
```diff
@@ -52,7 +53,9 @@ public OneInputOperatorWrapper(GeneratedClass<OneInputStreamOperator<IN, OUT>> g
 	public void setup(StreamTask<?, ?> containingTask, StreamConfig config,
 			Output<StreamRecord<OUT>> output) {
 		operator = generatedClass.newInstance(containingTask.getUserCodeClassLoader());
-		operator.setup(containingTask, config, output);
+		if (operator instanceof SetupableStreamOperator) {
+			((SetupableStreamOperator) operator).setup(containingTask, config, output);
```
Comment (pnowojski) [inline flink-table/flink-table-runtime-blink/src/main/java/org/apache/flink/table/runtime/OneInputOperatorWrapper.java]: Would it be viable to migrate this already to the non setupable operator? So to replace this if check with:
```
operator = generatedClass.newInstance(containingTask.getUserCodeClassLoader(), containingTask, config, output);
```
?

If it's non trivial it could be done as some kind of follow up step.

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/api/operators/SetupableStreamOperator.java [hunk 10]
```diff
@@ -0,0 +1,37 @@
...
+import org.apache.flink.streaming.runtime.streamrecord.StreamRecord;
+import org.apache.flink.streaming.runtime.tasks.StreamTask;
+
+/**
+ * Stream operators can implement this interface if they need access to the context and the output.
+ *
+ * @param <OUT> The output type of the operator
+ */
+@PublicEvolving
+public interface SetupableStreamOperator<OUT> {
```
Comment (pnowojski) [inline flink-streaming-java/src/main/java/org/apache/flink/streaming/api/operators/SetupableStreamOperator.java]: Mark the class `@Deprecated` and add java doc
> This class is deprecated in favour of using `StreamOperatorFactory` and it's `StreamOperatorFactory#createStreamOperator` and passing the required parameters to the `Operator`'s constructor in create method.

?

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/api/graph/JSONGenerator.java (hunk 4 above)
Comment (JingsongLi) [reply flink-streaming-java/src/main/java/org/apache/flink/streaming/api/graph/JSONGenerator.java]: Yes, new pr: https://github.com/apache/flink/pull/8343

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/api/operators/InputFormatOperatorFactory.java [hunk 11]
```diff
@@ -0,0 +1,41 @@
...
+import org.apache.flink.api.common.io.InputFormat;
+import org.apache.flink.core.io.InputSplit;
+import org.apache.flink.streaming.api.functions.source.InputFormatSourceFunction;
+
+/**
+ * Input format source operator factory which just wrap existed {@link StreamSource}.
+ *
+ * @param <OUT> The output type of the operator
+ */
+public class InputFormatOperatorFactory<OUT> extends SimpleOperatorFactory<OUT> {
```
Comment (pnowojski) [inline flink-streaming-java/src/main/java/org/apache/flink/streaming/api/operators/InputFormatOperatorFactory.java]: I think the dependency here should be reverted.

`InputFormatOperatorFactory` should be an interface extending `StreamOperatorFactory`, while the concrete class `SimpleInputFormatOperatorFactory` (with the code below) would be a specialisation of the `SimpleOperatorFactory`:
```
public class SimpleInputFormatOperatorFactory<OUT> 
  extends SimpleOperatorFactory<OUT> 
  implements InputFormatOperatorFactory<OUT>
```
Otherwise we still didn't fix the original issue: only `SimpleOperatorFactory` can have input format or an UDF.

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/api/operators/UdfStreamOperatorFactory.java [hunk 12]
```diff
@@ -0,0 +1,38 @@
...
+ */
+
+package org.apache.flink.streaming.api.operators;
+
+/**
+ * Udf stream operator factory which just wrap existed {@link AbstractUdfStreamOperator}.
+ *
+ * @param <OUT> The output type of the operator
+ */
+public class UdfStreamOperatorFactory<OUT> extends SimpleOperatorFactory<OUT> {
```
Comment (pnowojski) [inline flink-streaming-java/src/main/java/org/apache/flink/streaming/api/operators/UdfStreamOperatorFactory.java]: ditto about reversing the dependency

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/api/graph/StreamGraphHasherV2.java [hunk 13]
```diff
@@ -253,12 +252,9 @@ private boolean generateNodeHash(
 
 		if (LOG.isDebugEnabled()) {
 			String udfClassName = "";
-			if (node.getOperatorFactory() instanceof SimpleOperatorFactory) {
-				SimpleOperatorFactory factory = (SimpleOperatorFactory) node.getOperatorFactory();
-				if (factory.getOperator() instanceof AbstractUdfStreamOperator) {
-					udfClassName = ((AbstractUdfStreamOperator<?, ?>) factory.getOperator())
+			if (node.getOperatorFactory() instanceof UdfStreamOperatorFactory) {
+					udfClassName = ((UdfStreamOperatorFactory) node.getOperatorFactory()).getOperator()
```
Comment (pnowojski) [inline flink-streaming-java/src/main/java/org/apache/flink/streaming/api/graph/StreamGraphHasherV2.java]: Can not we add `getUserFunction()` or `getUserFunctionName()` method to the `UdfStreamOperatorFactory` interface?

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/api/graph/StreamingJobGraphGenerator.java [hunk 14]
```diff
@@ -649,14 +648,11 @@ private void configureCheckpointing() {
...
-			if (node.getOperatorFactory() instanceof SimpleOperatorFactory) {
-				SimpleOperatorFactory factory = (SimpleOperatorFactory) node.getOperatorFactory();
-				if (factory.getOperator() instanceof AbstractUdfStreamOperator) {
-					Function f = ((AbstractUdfStreamOperator<?, ?>) factory.getOperator()).getUserFunction();
-
-					if (f instanceof WithMasterCheckpointHook) {
-						hooks.add(new FunctionMasterCheckpointHookFactory((WithMasterCheckpointHook<?>) f));
-					}
+			if (node.getOperatorFactory() instanceof UdfStreamOperatorFactory) {
+				Function f = ((UdfStreamOperatorFactory) node.getOperatorFactory()).getOperator().getUserFunction();
```
Comment (pnowojski) [inline flink-streaming-java/src/main/java/org/apache/flink/streaming/api/graph/StreamingJobGraphGenerator.java]: Can not we add `getUserFunction()` method to the `UdfStreamOperatorFactory` interface?

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/api/graph/StreamGraphHasherV2.java [hunk 15]
```diff
@@ -285,8 +284,8 @@ private boolean isChainable(StreamEdge edge, boolean isChainingEnabled, StreamGr
 		StreamNode upStreamVertex = streamGraph.getSourceVertex(edge);
 		StreamNode downStreamVertex = streamGraph.getTargetVertex(edge);
 
-		StreamOperator<?> headOperator = upStreamVertex.getOperator();
-		StreamOperator<?> outOperator = downStreamVertex.getOperator();
+		StreamOperatorFactory<?> headOperator = upStreamVertex.getOperatorFactory();
```
Comment (aljoscha) [inline flink-streaming-java/src/main/java/org/apache/flink/streaming/api/graph/StreamGraphHasherV2.java]: Is this compatible with previous savepoints? Because we now hash the factory not the operator.

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/api/operators/StreamOperatorFactory.java (hunk 8 above)
Comment (aljoscha) [reply flink-streaming-java/src/main/java/org/apache/flink/streaming/api/operators/StreamOperatorFactory.java]: I'm thinking `@Internal`

**Comment by aljoscha** [comment]
From a design perspective this change looks good. 👌  I had some comments inline.

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/api/graph/StreamGraphHasherV2.java (hunk 15 above)
Comment (JingsongLi) [reply flink-streaming-java/src/main/java/org/apache/flink/streaming/api/graph/StreamGraphHasherV2.java]: The hash value should not be changed:
The hash is either computed from the transformation's user-specified id or generated in a deterministic way. user-specified id not be changed.
The generated hash is deterministic with respect to:
1.node-local properties (node ID): node ID is generated by id counter.
2.chained output nodes: Chaining strategy not be changed.
3.input nodes hashes

So I think not related to StreamOperatorFactory.

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/api/operators/StreamOperatorFactory.java (hunk 8 above)
Comment (JingsongLi) [reply flink-streaming-java/src/main/java/org/apache/flink/streaming/api/operators/StreamOperatorFactory.java]: OK done
